baseapp/baseapp_test.go

```diff
@@ -2153,3 +2153,122 @@ func TestBaseApp_Init(t *testing.T) {
+		Candidate Answer: For now, Block is just a list of a transactions and a header with an app hash
+
+		Note: In the current implementation, all substores might not be SMTs, do we assume they are?
+
+	*/
+
+	codec := codec.NewLegacyAmino()
+	registerTestCodec(codec)
+
+	storeTraceBuf := &bytes.Buffer{}
+	subStoreTraceBuf := &bytes.Buffer{}
+
+	routerOpt := func(bapp *BaseApp) {
+		bapp.Router().AddRoute(sdk.NewRoute(routeMsgKeyValue, func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+			kv := msg.(*msgKeyValue)
+			bapp.cms.GetKVStore(capKey2).Set(kv.Key, kv.Value)
+			return &sdk.Result{}, nil
+		}))
+	}
+
+	app := setupBaseApp(t,
+		AppOptionFunc(routerOpt),
+		SetSubstoreTracer(storeTraceBuf),
+		SetTracerFor(capKey2, subStoreTraceBuf),
+	)
+
+	app.InitChain(abci.RequestInitChain{})
+
+	blocks := 1
+	txsPerBlock := 2
+
+	r := rand.New(rand.NewSource(3920758213583))
+	keyCounter := 0
+	for height := int64(1); height <= int64(blocks); height++ {
+		app.BeginBlock(abci.RequestBeginBlock{Header: tmproto.Header{Height: height}})
+
+		for txNum := 0; txNum < txsPerBlock; txNum++ {
+			tx := txTest{Msgs: []sdk.Msg{}}
+			for msgNum := 0; msgNum < 2; msgNum++ {
+				key := []byte(fmt.Sprintf("%v", keyCounter))
+				value := make([]byte, 10000)
+				_, err := r.Read(value)
+				require.NoError(t, err)
+				tx.Msgs = append(tx.Msgs, msgKeyValue{Key: key, Value: value})
+				keyCounter++
+			}
+			txBytes, err := codec.Marshal(tx)
+			require.NoError(t, err)
+			resp := app.DeliverTx(abci.RequestDeliverTx{Tx: txBytes})
+			require.True(t, resp.IsOK(), "%v", resp.String())
+		}
+		app.EndBlock(abci.RequestEndBlock{Height: height})
+		app.Commit()
+
+		cms := app.cms.(*multi.Store)
+
+		var fraudproof FraudProof
+		fraudproof.blockHeight = uint64(app.LastBlockHeight())
+
+		// Go over all storeKeys inside app.cms and populate values inside fraudproof
+		storeKeys := cms.GetAllStoreKeys()
+		for storeKey := range storeKeys {
+			// Generates a deepsubtree for substore with given key
+			kvStore := cms.GetKVStore(storeKey)
+			traceKv := kvStore.(*tracekv.Store)
+			keys := traceKv.GetAllKeysUsedInTrace(*subStoreTraceBuf)
+
+			substoreSMT := app.cms.(*multi.Store).GetSubStoreSMT(storeKey.Name())
+
+			root := substoreSMT.Root()
+
+			var stateWitness StateWitness
+			stateWitness.root = root
+
+			// deepsubtree := smt.NewDeepSparseMerkleSubTree(smt.NewSimpleMap(), smt.NewSimpleMap(), sha512.New512_256(), root)
+			for key := range keys {
+				var witnessData WitnessData
+				value := substoreSMT.Get([]byte(key))
+				proof, err := substoreSMT.GetSMTProof([]byte(key))
+				require.Nil(t, err)
+				// deepsubtree.AddBranch(*proof, []byte(key), []byte(value))
+				witnessData.Key = []byte(key)
+				witnessData.Value = []byte(value)
+				witnessData.proof = *proof
+				stateWitness.WitnessData = append(stateWitness.WitnessData, witnessData)
+			}
+			fraudproof.stateWitness[storeKey.Name()] = stateWitness
```

Comment:
I assume this will be refactored into a generateFraudProof method later?

Reply:
yes indeed